# Human-Genome_assembly_review

## Papers for Review
##### Papers About Reference Genome
1. [Is it time to change the reference genome?](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1774-4)
2. [Evaluation of GRCh38 and de novo haploid genome assemblies demonstrates the enduring quality of the reference assembly](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5411779/)

Blogs
1. [Reference Genome Improvement](https://www.genome.wustl.edu/items/reference-genome-improvement/)
2. [A critical tool for human genomics and precision medicine: De novo human genome assembly](https://research.wisc.edu/a-critical-tool-for-human-genomics-and-precision-medicine-de-novo-human-genome-assembly/)
3. [New High-Resolution Genome Assemblies Expand Our Understanding of Human-Ape Differences](https://www.pacb.com/blog/new-high-resolution-genome-assemblies-expand-our-understanding-of-human-ape-differences/) 
4. [Genome Reference Consortium](https://www.sanger.ac.uk/data/genome-reference-consortium/)
5. Blog by [Andrew Severin](https://bioinformaticsworkbook.org/dataAnalysis/GenomeAssembly/Intro_GenomeAssembly.html#gsc.tab=0) and sequencing [basics](https://www.healio.com/hematology-oncology/learn-genomics/whole-genome-sequencing/whole-genome-sequencing-methods)
6. [Writing the human genome](https://cen.acs.org/articles/95/i28/Writing-human-genome.html)
7. [Dawn of good, fast and cheap human genome assembly](https://www.sciencedaily.com/releases/2020/05/200504114131.htm)
8. [Lessons from the Human Genome Project](http://sitn.hms.harvard.edu/flash/2019/lessons-from-the-human-genome-project/)
9. [Nanopore Blogs](https://nanoporetech.com/taxonomy/term/258)







##### Papers About Reference Genome Assembly

1.  [Human Genome Assembly in 100 Minutes](https://www.biorxiv.org/content/10.1101/705616v1.full)
2. Assembler 1: [SAGE2: parallel human genome assembly](https://academic.oup.com/bioinformatics/article/34/4/678/4553153)
3. [Nanopore sequencing and assembly of a human genome with ultra-long reads](https://www.nature.com/articles/nbt.4060)
4. [Dense and accurate whole-chromosome haplotyping of individual genomes.](https://europepmc.org/article/pmc/pmc5670131#free-full-text)
5. [The Theory and Practice of Genome Sequence Assembly](https://pubmed.ncbi.nlm.nih.gov/25939056/)
6. [NovoGraph: Human genome graph construction from multiple long-read de novo assemblies](https://f1000research.com/articles/7-1391)
7. [Genetic variation and the de novo assembly of human genomes](https://www.semanticscholar.org/paper/Genetic-variation-and-the-de-novo-assembly-of-human-Chaisson-Wilson/f5cfb387a6bd8a1cb7bdf07562fadb9a351e37cd)
8. [Ten steps to get started in Genome Assembly and Annotation](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5850084/)
9. [The A, C, G, and T of Genome Assembly](https://www.hindawi.com/journals/bmri/2016/6329217/)
10. [The bioinformatics tools for the genome assembly and analysis based on third-generation sequencing](https://academic.oup.com/bfg/article-abstract/18/1/1/5181442?redirectedFrom=PDF) 
11. [Current challenges and solutions of de novo assembly](https://link.springer.com/article/10.1007/s40484-019-0166-9)
12. [Reference-guided de novo assembly approach improves genome reconstruction for related species](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-017-1911-6)
13. [A field guide to whole‐genome sequencing, assembly and annotation](https://onlinelibrary.wiley.com/doi/full/10.1111/eva.12178)
14. [New Approaches for Genome Assembly and Scaffolding](https://pubmed.ncbi.nlm.nih.gov/30485757/)
15. [Bioinformatics and Computational Tools for Next-Generation Sequencing Analysis in Clinical Genetics](https://www.mdpi.com/2077-0383/9/1/132)
16. [Is It Ordered Correctly? Validating Genome Assemblies by Optical Mapping](http://www.plantcell.org/content/30/1/7)
17. [Improving ancient DNA genome assembly](https://peerj.com/articles/3126/)
18. [GAGE: A critical evaluation of genome assemblies and assembly algorithms](http://m.genome.cshlp.org/content/genome/22/3/557.full.html)
19. [GRASShopPER—An algorithm for de novo assembly based on GPU alignments](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0202355)
20. [Using Apache Spark on genome assembly for scalable overlap-graph reduction](https://humgenomics.biomedcentral.com/articles/10.1186/s40246-019-0227-1)
21. [De novo assembly using Illumina Reads](https://www.illumina.com/Documents/products/technotes/technote_denovo_assembly_ecoli.pdf)
22. [SORA: Scalable Overlap-graph Reduction Algorithms for Genome Assembly using Apache Spark in the Cloud](https://ieeexplore.ieee.org/document/8621546)
23. [Versatile genome assembly evaluation with QUAST-LG](https://academic.oup.com/bioinformatics/article/34/13/i142/5045727) 
24. [Meraculous2: fast accurate short-read assembly of large polymorphic genomes](Meraculous2:%20fast%20accurate%20short-read%20assembly%20of%20large%20polymorphic%20genomes)
25. [MitoZ: a toolkit for animal mitochondrial genome assembly, annotation and visualization](https://academic.oup.com/nar/article/47/11/e63/5377471)
26. [Norgal: extraction and de novo assembly of mitochondrial DNA from whole-genome sequencing data](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-017-1927-y)
27. [A simple method for sequencing the whole human mitochondrial genome directly from samples and its application to genetic testing](https://www.nature.com/articles/s41598-019-53449-y)
28. [Investigating Human Mitochondrial Genomes in Single Cells](https://www.mdpi.com/2073-4425/11/5/534)
29. [Next-Generation Sequencing of Human Mitochondrial Reference Genomes Uncovers High Heteroplasmy Frequency](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002737)
30. [SMART: Statistical Mitogenome Assembly with Repeats](https://www.biorxiv.org/content/10.1101/795633v1)
31. [Statistical Mitogenome Assembly with RepeaTs](https://www.liebertpub.com/doi/abs/10.1089/cmb.2019.0505?journalCode=cmb&)
32. [MitoSuite: a graphical tool for human mitochondrial genome profiling in massive parallel sequencing](https://peerj.com/articles/3406/)
33. [Reliable Resequencing of Human Mitochondrial Genome to Identify Mutations Linked to Mitochondrial Related Diseases](https://assets.thermofisher.com/TFS-Assets/LSG/brochures/cms_053271.pdf)
34. [mitoMaker: A Pipeline for Automatic Assembly and Annotation of Animal Mitochondria Using Raw NGS Data](https://www.preprints.org/manuscript/201808.0423/v1)
35. [GMASS: a novel measure for genome assembly structural similarity](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-019-2710-z#:~:text=Conclusions-,The%20GMASS%20score%20is%20a%20novel%20measure%20for%20representing%20structural,and%20developing%20de%20novo%20assemblers.)
36. [dnAQET: a framework to compute a consolidated metric for benchmarking quality of de novo assemblies](https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-019-6070-x)
37. [CONNET: Accurate Genome Consensus in Assembling Nanopore Sequencing Data via Deep Learning](https://www.sciencedirect.com/science/article/pii/S2589004220303138)
38. [Assembly of long, error-prone reads using repeat graphs](https://www.nature.com/articles/s41587-019-0072-8)
39. [Scalable Genome Assembly through Parallel de Bruijn Graph Construction for Multiple k-mers](https://www.nature.com/articles/s41598-019-51284-9)
40. [GPU-Accelerated Large-Scale Genome Assembly](https://ieeexplore.ieee.org/document/8425235)
41. [BlobToolKit – Interactive Quality Assessment of Genome Assemblies](https://www.g3journal.org/content/10/4/1361)
42. [Minimap2: pairwise alignment for nucleotide sequences](https://academic.oup.com/bioinformatics/article/34/18/3094/4994778)
43. [Graph-based genome alignment and genotyping with HISAT2 and HISAT-genotype](https://www.nature.com/articles/s41587-019-0201-4)
44. [Faucet: streaming de novo assembly graph construction](https://academic.oup.com/bioinformatics/article/34/1/147/4004871) 
